src/assets/TokenListController.ts

```diff
@@ -0,0 +1,270 @@
+  tokenList: {},
+  tokensChainsCache: {},
+};
+
+/**
+ * Controller that passively polls on a set interval for the list of tokens from metaswaps api
+ */
+export class TokenListController extends BaseController<
+  typeof name,
+  TokenListState
+> {
+  private mutex = new Mutex();
+
+  private intervalId?: NodeJS.Timeout;
+
+  private intervalDelay: number;
+
+  private cacheRefreshThreshold: number;
+
+  private chainId: string;
+
+  /**
+   * Creates a TokenListController instance
+   *
+   * @param options - Constructor options
+   * @param options.interval - The polling interval, in milliseconds
+   * @param options.messenger - A reference to the messaging system
+   * @param options.state - Initial state to set on this controller
+   */
+  constructor({
+    chainId,
+    onNetworkStateChange,
+    interval = DEFAULT_INTERVAL,
+    cacheRefreshThreshold = DEFAULT_THRESHOLD,
+    messenger,
+    state,
+  }: {
+    chainId: string;
+    onNetworkStateChange: (
+      listener: (networkState: NetworkState) => void,
+    ) => void;
+    interval?: number;
+    cacheRefreshThreshold?: number;
+    messenger: RestrictedControllerMessenger<
+      typeof name,
+      GetTokenListState,
+      TokenListStateChange,
+      never,
+      never
+    >;
+    state?: Partial<TokenListState>;
+  }) {
+    super({
+      name,
+      metadata,
+      messenger,
+      state: { ...defaultState, ...state },
+    });
+    this.intervalDelay = interval;
+    this.cacheRefreshThreshold = cacheRefreshThreshold;
+    this.chainId = chainId;
+    onNetworkStateChange((networkState) => {
+      this.chainId = networkState.provider.chainId;
+    });
+  }
+
+  /**
+   * Start polling for the token list
+   */
+  async start() {
+    await this.startPolling();
+  }
+
+  /**
+   * Stop polling for the token list
+   */
+  stop() {
+    this.stopPolling();
+  }
+
+  /**
+   * Prepare to discard this controller.
+   *
+   * This stops any active polling.
+   */
+  destroy() {
+    super.destroy();
+    this.stopPolling();
+  }
+
+  private stopPolling() {
+    if (this.intervalId) {
+      clearInterval(this.intervalId);
+    }
+  }
+
+  /**
+   * Starts a new polling interval
+   */
+  private async startPolling(): Promise<void> {
+    await safelyExecute(() => this.fetchTokenList());
+    this.intervalId = setInterval(async () => {
+      await safelyExecute(() => this.fetchTokenList());
+    }, this.intervalDelay);
+  }
+
+  /**
+   * Fetching token list from the Token Service API
+   */
+  async fetchTokenList(): Promise<void> {
+    const releaseLock = await this.mutex.acquire();
+    try {
+      const tokensFromAPI: Token[] = await safelyExecute(() =>
+        this.fetchFromCache(),
+      );
+      const { tokensChainsCache } = this.state;
+      const tokenList: TokenMap = {};
+
+      // filtering out tokens with less than 2 occurences
+      const filteredTokenList = tokensFromAPI.filter(
+        (token) => token.occurrences >= 2,
```

Comment:
Just wondering, how did we determine this number of occurrences to filter by?

Reply:
The least required occurrences are determined as 2 based on the assumption that it is less likely that a malicious/scammy or spammy token gets listed in 2 token lists by a known/well-established team